Refactor Daemon Identity allocation routines

joestringer opened the pull request:
See commit messages for more details. No functional changes intended.

Used to implement fix https://github.com/cilium/cilium/pull/17699 .

joestringer committed:
testutils: Shift identity mocking to dedicated dir

The mock identity notifier and identity allocator make sense together
under pkg/testutils/identity, and refactoring like this will be required
by upcoming commits in order to avoid circular imports.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
testutils: Add NewFakeIdentityAllocator contructor

This constructor will simplify upcoming commits by performing all of the
necessary initialization for the FakeIdentityAllocator in all tests that
make use of this structure.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache: Rename ReleaseCIDRs{,IdentitiesByCIDR}

An upcoming commit will introduce a new variation on the release
function, so just rename this one here to simplify the upcoming changes.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
daemon: Abstract identity allocation from FQDN

In order to test an upcoming bugfix, the internal FQDN handling
functions that are triggered from the unit tests must be able to refer
to an abstract identity allocator. This will then be mocked out by the
tests.

This commit extends the existing cache.IdentityAllocator interface to
support allocating identities for CIDRs. A future commit will also add
Identity release functions to the same abstraction.

This also paves the way for some improvements to code structure: Some of
the FQDN logic under `daemon/cmd/fqdn.go` may be able to be pushed down
into pkg/fqdn now that it doesn't directly reference the globals from
the ipcache package.

There's another remaining open question around code structure for how
the IPcache and identity allocation packages handle the pattern for
allocating CIDR identities vs. injecting those identities into the
underlying IPcache maps (eg in eBPF). This commit shies away from
solving that problem right now in order to keep the changes minimal as
this code is likely to be proposed for backport.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
policy: Initialize repository with idAllocator

Upcoming commits will rely on the daemon passing the identityAllocator
module down to the fqdn subsystem via the policy repository. For now,
adjust the function signatures to pass this through, but avoid actually
using the new parameter for now.

This commit is a functional no-op but reduces the noise in terms of code
churn from upcoming commits.

Signed-off-by: Joe Stringer <joe@cilium.io>

christarazi commented on daemon/cmd/daemon.go:
```diff
@@ -832,7 +832,8 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 		// Ignore the channel returned by this function, as we want the global
 		// identity allocator to run asynchronously.
-		d.identityAllocator.InitIdentityAllocator(k8s.CiliumClient(), nil)
+		realIdentityAllocator := d.identityAllocator.(cachingIdentityAllocator)
```
Having to type assert the `identityAllocator` everywhere seems a bit awkward. But given that the call we want to abstract (`AllocateCIDRsForIPs`) is not a method on a type, just a function, then I can see why this approach was taken. Can't think of alternatives at the moment, but would be nice to avoid the type assertions.

joestringer replied:
IIRC the reason I did this wasn't anything to do with the existing `IdentityAllocator` interface, but more that `CachingIdentityAllocator` has these initialization routines and so on that didn't seem to generically make sense on the `IdentityAllocator` type. Maybe it's enough to just declare a new interface in `pkg/daemon` that embeds the `IdentityAllocator`.  It could have these extra functions so that the daemon can initialize and use the `CachingIdentityAllocator` type via an abstract representation, but still pass an abstract `IdentityAllocator` down to child calls which don't need to worry about initialization. Hopefully this will also minimize the amount of boilerplate required to implement these in places like the testsuite.

joestringer replied:
^^ I did this, and it came out quite clean. :+1:

joestringer committed:
identity: Add CIDR release function to allocator

Everything that goes up must eventually come down. And so it is also
with everything that is allocated; it must be deallocated.

Extend the IdentityAllocator interface to also support CIDR release
functions and plumb through an initial implementation. This is currently
not hooked into the agent itself, this will come with an upcoming bugfix
instead.

Signed-off-by: Joe Stringer <joe@cilium.io>

christarazi reviewed (approved):
:rocket:

ldelossa reviewed (approved):
Nice refactor :+1:

jrajahalme reviewed (approved):
LGTM
